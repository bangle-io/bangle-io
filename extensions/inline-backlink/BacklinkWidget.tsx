import React, { useCallback, useEffect, useMemo, useState } from 'react';

import { nsmApi2 } from '@bangle.io/api';
import { isAbortError } from '@bangle.io/mini-js-utils';
import type { HighlightTextType, SearchMatch } from '@bangle.io/search-pm-node';
import {
  ButtonIcon,
  ChevronDownIcon,
  ChevronRightIcon,
  Sidebar,
} from '@bangle.io/ui-components';
import {
  cx,
  getMouseClickType,
  MouseClick,
  safeCancelIdleCallback,
  safeRequestIdleCallback,
} from '@bangle.io/utils';
import { naukarProxy } from '@bangle.io/worker-naukar-proxy';
import { createWsPath, removeExtension, resolvePath } from '@bangle.io/ws-path';

const IconStyle = {
  height: 16,
  width: 16,
};

interface BacklinkSearchResult {
  wsPath: string;
  matches: SearchMatch[];
}

export function BacklinkWidget() {
  const backlinkSearchResult = useBacklinkSearch();

  const makeOnClick = useCallback((_path?: string) => {
    return (event: React.MouseEvent<any>) => {
      if (!_path) {
        return;
      }

      const wsPath = createWsPath(_path);

      const clickType = getMouseClickType(event);

      if (clickType === MouseClick.NewTab) {
        nsmApi2.workspace.openWsPathInNewTab(wsPath);
      } else if (clickType === MouseClick.ShiftClick) {
        nsmApi2.workspace.pushSecondaryWsPath(wsPath);
      } else {
        nsmApi2.workspace.pushPrimaryWsPath(wsPath);
      }
    };
  }, []);

  const [openedItems, updateOpenedItems] = useState(() => new Set<string>());
  const isCollapsed = useCallback(
    (r: BacklinkSearchResult) => {
      return !openedItems.has(r.wsPath);
    },
    [openedItems],
  );

  return (
    <div
      data-testid="inline-backlink_widget-container"
      className="flex flex-col"
    >
      {!backlinkSearchResult || backlinkSearchResult.length === 0 ? (
        <span>
          üêí No backlinks found!
          <br />
          <span className="font-light">
            Create one by typing <kbd className="font-normal">[[</kbd> followed
            by the name of the note.
          </span>
        </span>
      ) : (
        backlinkSearchResult.map((r, i) => {
          return (
            <React.Fragment key={i}>
              <Sidebar.Row2
                titleClassName=""
                className={'rounded text-sm truncate py-1 select-none'}
                extraInfoClassName="ml-1 text-sm"
                onClick={makeOnClick(r.wsPath)}
                item={{
                  uid: r.wsPath,
                  showDividerAbove: false,
                  title: resolvePath(r.wsPath).fileNameWithoutExt,
                  leftNode: (
                    <ButtonIcon
                      onClick={(e) => {
                        updateOpenedItems((items) => {
                          if (items.has(r.wsPath)) {
                            const clone = new Set(items);
                            clone.delete(r.wsPath);

                            return clone;
                          } else {
                            const clone = new Set(items);
                            clone.add(r.wsPath);

                            return clone;
                          }
                        });
                        e.stopPropagation();
                        e.preventDefault();
                      }}
                    >
                      {isCollapsed(r) ? (
                        <ChevronRightIcon style={IconStyle} />
                      ) : (
                        <ChevronDownIcon style={IconStyle} />
                      )}
                    </ButtonIcon>
                  ),
                  rightNode: (
                    <ButtonIcon className="text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1">
                      {r.matches.length}
                    </ButtonIcon>
                  ),
                  rightHoverNode: (
                    <ButtonIcon className="text-xs font-semibold rounded bg-colorNeutralSolidFaint px-1">
                      {r.matches.length}
                    </ButtonIcon>
                  ),
                }}
              />
              {!isCollapsed(r) &&
                r.matches.map((matchObj, j) => (
                  <Sidebar.Row2
                    key={j}
                    className={
                      'B-search-notes_search-result-text-match ml-1 pl-3 rounded ' +
                      (j === 0 ? 'mt-0' : 'mt-1')
                    }
                    onClick={makeOnClick(r.wsPath)}
                    titleClassName="text-sm"
                    item={{
                      uid: 'search-result-text-match-' + j,
                      title: <HighlightText highlightText={matchObj.match} />,
                    }}
                  ></Sidebar.Row2>
                ))}
            </React.Fragment>
          );
        })
      )}
    </div>
  );
}

function useBacklinkSearch(): BacklinkSearchResult[] | undefined {
  const { focusedEditorId } = nsmApi2.editor.useEditor();
  const { wsName, openedWsPaths } = nsmApi2.workspace.useWorkspace();

  const [results, updateResults] = useState<BacklinkSearchResult[] | undefined>(
    undefined,
  );

  const calculateResult = useCallback(
    (focusedWsPath: string, controller: AbortController) => {
      if (!wsName) {
        return;
      }
      naukarProxy.abortable
        .abortableSearchWsForPmNode(controller.signal, wsName, 'backlink:*', [
          {
            nodeName: 'wikiLink',
            dataAttrName: 'path',
            printBefore: '[[',
            printAfter: ']]',
            queryIdentifier: 'backlink:',
          },
        ])
        .then(
          (result) => {
            const fileName = removeExtension(
              resolvePath(focusedWsPath).fileName,
            );

            updateResults(
              result
                .map((r) => {
                  const newMatches = r.matches.filter((match) => {
                    const [, highlightTextMatch] = match.match;

                    if (highlightTextMatch) {
                      return highlightTextMatch.includes(fileName);
                    }

                    return false;
                  });

                  return {
                    wsPath: r.uid,
                    matches: newMatches,
                  };
                })
                .filter((r) => r.matches.length > 0),
            );
          },
          (error) => {
            if (isAbortError(error)) {
              return;
            }

            throw error;
          },
        );
    },
    [wsName],
  );

  const focusedWsPath = useMemo(() => {
    if (focusedEditorId != null) {
      return openedWsPaths.getByIndex(focusedEditorId);
    }

    return undefined;
  }, [openedWsPaths, focusedEditorId]);

  useEffect(() => {
    const controller = new AbortController();
    let cb =
      focusedWsPath &&
      safeRequestIdleCallback(
        () => {
          calculateResult(focusedWsPath, controller);
        },
        { timeout: 1000 },
      );

    return () => {
      controller.abort();

      if (cb) {
        safeCancelIdleCallback(cb);
      }
    };
  }, [focusedWsPath, wsName, calculateResult]);

  return results;
}

function HighlightText({
  highlightText,
}: {
  highlightText: HighlightTextType;
}) {
  return (
    <div className="B-inline-backlink_highlight-text-container">
      {highlightText.map((t, i) => (
        <span
          key={i}
          className={cx(i % 2 === 1 && 'bg-colorPromoteSolidFaint', 'text-sm')}
        >
          {t}
        </span>
      ))}
    </div>
  );
}
