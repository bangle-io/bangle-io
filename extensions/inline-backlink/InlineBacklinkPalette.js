import React, { useMemo } from 'react';
import reactDOM from 'react-dom';
import {
  useInlinePaletteItems,
  useInlinePaletteQuery,
} from 'inline-palette/index';
import { SidebarRow } from 'ui-components';
import { replaceSuggestionMarkWith } from 'inline-palette/inline-palette';
import { resolvePath, useGetCachedWorkspaceFiles } from 'workspace/index';
import { backLinkNodeName, palettePluginKey } from './config';

export function InlineBacklinkPalette() {
  const { query, counter } = useInlinePaletteQuery(palettePluginKey);
  const [currentFiles = []] = useGetCachedWorkspaceFiles();

  const items = useMemo(() => {
    if (!query) {
      return [];
    }

    const items = currentFiles.map((wsPath) => {
      return {
        uid: wsPath,
        title: wsPath,
        editorExecuteCommand: ({ item }) => {
          return (state, dispatch, view) => {
            let nodeType = state.schema.nodes[backLinkNodeName];
            return replaceSuggestionMarkWith(
              palettePluginKey,
              nodeType.create({
                title: resolvePath(wsPath).fileName,
                wsPath: wsPath,
              }),
            )(state, dispatch, view);
          };
        },
      };
    });
    return items.filter((item) => queryMatch(item, query)).slice(0, 50);
  }, [query, currentFiles]);

  const { tooltipContentDOM, getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
  );

  return reactDOM.createPortal(
    <div className="bangle-emoji-suggest">
      {query === '' && (
        <SidebarRow
          dataId="searchNote"
          className="palette-row"
          title="Search for a note"
        />
      )}
      {items.map((item, i) => {
        return (
          <SidebarRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            title={item.title}
            rightHoverIcon={item.rightHoverIcon}
            rightIcon={
              <kbd className="whitespace-nowrap">{item.keybinding}</kbd>
            }
            {...getItemProps(item, i)}
          />
        );
      })}
    </div>,
    tooltipContentDOM,
  );
}

function queryMatch(item, query) {
  const keywords = item.keywords || '';

  if (keywords.length > 0) {
    if (strMatch(keywords.split(','), query)) {
      return item;
    }
  }
  return strMatch(item.title, query) ? item : undefined;
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase();
  return a.includes(b) || b.includes(a);
}
