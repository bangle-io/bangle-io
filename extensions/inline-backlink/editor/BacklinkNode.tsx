import React, { useEffect, useState } from 'react';

import type { EditorState } from '@bangle.dev/pm';

import { nsmApi2 } from '@bangle.io/api';
import { EditorDisplayType } from '@bangle.io/constants';
import { getEditorPluginMetadata } from '@bangle.io/editor-common';
import type { RenderReactNodeView } from '@bangle.io/extension-registry';
import type { WsPath } from '@bangle.io/shared-types';
import { useHover, useTooltipPositioner } from '@bangle.io/ui-components';

import { BacklinkNodeButton } from '../BacklinkNodeButton';
import { calcWikiLinkMapping, getAllWikiLinks } from '../calculate-matches';
import { backlinkNodeName } from '../config';
import { LinkPreview } from './LinkPreview';
import { useOnClickBacklink } from './use-on-click-backlink';

export function BacklinkNode({
  nodeAttrs,
  editorState,
}: {
  nodeAttrs: { path: string; title?: string };
  editorState: EditorState;
}) {
  const { wsPath: currentWsPath, editorDisplayType } =
    getEditorPluginMetadata(editorState);

  const { wsName, noteWsPaths } = nsmApi2.workspace.useWorkspace();

  let { path: wikiLink, title } = nodeAttrs;
  title = title || wikiLink;

  const onClick = useOnClickBacklink({
    wikiLink,
    currentWsPath,
    editorState,
  });

  const [backlinksWsPath, updateBacklinksWsPath] = useState<WsPath | undefined>(
    undefined,
  );

  const disablePopup = editorDisplayType === EditorDisplayType.Floating;

  const [isHovered, updateIsHovered] = useState(false);
  const { hoverProps: tooltipHoverProps, isHovered: isTooltipHovered } =
    useHover({ isDisabled: disablePopup });

  const {
    isTooltipVisible,
    setTooltipElement,
    setTriggerElement,
    tooltipProps,
  } = useTooltipPositioner({
    isDisabled: disablePopup,
    isActive: !disablePopup && (isHovered || isTooltipHovered),
    xOffset: 10,
    yOffset: 0,
    // TODO we can optimize where we position based on where empty
    // space exists
    placement: 'right',
    delay: 450,
    immediateClose: false,
  });

  useEffect(() => {
    if (noteWsPaths) {
      const result = calcWikiLinkMapping(
        noteWsPaths,
        getAllWikiLinks(editorState),
      ).get(wikiLink);

      updateBacklinksWsPath(result);
    }
  }, [wsName, noteWsPaths, wikiLink, editorState]);

  useEffect(() => {
    let targetWsPath = backlinksWsPath;
    let didSet = false;

    if (targetWsPath && isTooltipVisible) {
      didSet = true;
      nsmApi2.workspace.pushOpenedWsPath((openedWsPaths) => {
        return openedWsPaths.updatePopupEditorWsPath(targetWsPath);
      });
    }

    return () => {
      if (didSet) {
        nsmApi2.workspace.pushOpenedWsPath((openedWsPaths) => {
          // make sure we are unsetting the correct wsPath
          if (
            targetWsPath &&
            targetWsPath === openedWsPaths.popupEditorWsPath
          ) {
            return openedWsPaths.updatePopupEditorWsPath(undefined);
          }

          return openedWsPaths;
        });
      }
    };
  }, [backlinksWsPath, isTooltipVisible]);

  return (
    <>
      <BacklinkNodeButton
        ref={setTriggerElement}
        linkNotFound={!backlinksWsPath}
        title={title}
        onClick={onClick}
        onHoverChange={updateIsHovered}
      />
      {isTooltipVisible && backlinksWsPath && (
        <LinkPreview
          disablePreview={disablePopup}
          positionProps={{ ...tooltipHoverProps, ...tooltipProps.attributes }}
          ref={setTooltipElement}
          style={tooltipProps.style}
          wsPath={backlinksWsPath}
        />
      )}
    </>
  );
}

export const renderReactNodeView: RenderReactNodeView = {
  [backlinkNodeName]: ({ nodeViewRenderArg }) => {
    const { path, title } = nodeViewRenderArg.node.attrs;

    if (typeof path !== 'string') {
      return <span>Invalid Path</span>;
    }

    return (
      <BacklinkNode
        nodeAttrs={{ path, title }}
        editorState={nodeViewRenderArg.view.state}
      />
    );
  },
};
