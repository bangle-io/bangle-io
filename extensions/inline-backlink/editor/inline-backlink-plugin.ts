import { NodeView } from '@bangle.dev/core';

import { backlinkNodeName } from '../config';

export function inlineBacklinkPlugin() {
  return () => [
    NodeView.createPlugin({
      name: backlinkNodeName,
      // inline-block allows the span to get full height of image
      // or else folks depending on the boundingBox get incorrect
      // dimensions.
      containerDOM: ['span', { style: 'display: inline-block;' }],
    }),
  ];
}
