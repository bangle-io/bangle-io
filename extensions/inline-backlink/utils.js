export function removeMdExtension(str) {
  if (str.endsWith('.md')) {
    return str.slice(0, -3);
  }
  return str;
}
