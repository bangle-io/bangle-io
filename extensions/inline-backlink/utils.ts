import {
  filePathToWsPath,
  removeExtension,
  resolvePath,
  sanitizeFilePath,
  suffixWithNoteExtension,
  validateWsPath,
} from '@bangle.io/ws-path';

export function wsPathFromQuery(query: string, wsName: string) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = suffixWithNoteExtension(filePath);
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}

export function getBacklinkPath(wsPath: string, allWsPaths: readonly string[]) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeExtension(filePath);
  }

  return removeExtension(fileName);
}
