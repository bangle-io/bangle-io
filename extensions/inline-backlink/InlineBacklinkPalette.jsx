import React, { useMemo, useEffect, useContext } from 'react';
import reactDOM from 'react-dom';
import { useEditorViewContext } from '@bangle.dev/react';
import {
  useInlinePaletteItems,
  useInlinePaletteQuery,
  replaceSuggestionMarkWith,
} from 'inline-palette/index';
import { UniversalPalette, InlinePaletteRow } from 'ui-components/index';
import {
  resolvePath,
  filePathToWsPath,
  validateWsPath,
  sanitizeFilePath,
} from 'ws-path';
import { backLinkNodeName, palettePluginKey } from './config';
import { conditionalSuffix, removeMdExtension } from 'utils/index';
import { useWorkspaceContext } from 'workspace-context/index';

// Creating this also closes the palette
const createBackLinkNode = (wsPath, allNoteWsPaths) => {
  return (state, dispatch, view) => {
    const nodeType = state.schema.nodes[backLinkNodeName];
    const backLinkPath = getBacklinkPath(wsPath, allNoteWsPaths);

    return replaceSuggestionMarkWith(
      palettePluginKey,
      nodeType.create({
        path: backLinkPath,
      }),
    )(state, dispatch, view);
  };
};
export function InlineBacklinkPalette() {
  const { query, counter, tooltipContentDOM } =
    useInlinePaletteQuery(palettePluginKey);

  return reactDOM.createPortal(
    <div className="inline-palette-wrapper shadow-2xl">
      <div className="inline-palette-items-wrapper">
        {/* TODO I am unable to hide inner when palette is invisible */}
        <InlineBacklinkPaletteInner query={query} counter={counter} />
        <UniversalPalette.PaletteInfo>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">â†‘â†“</kbd> Navigate
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Enter</kbd> Create link
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Esc</kbd> Dismiss
          </UniversalPalette.PaletteInfoItem>
        </UniversalPalette.PaletteInfo>
      </div>
    </div>,
    tooltipContentDOM,
  );
}

function InlineBacklinkPaletteInner({ query, counter }) {
  const view = useEditorViewContext();
  const { wsName, noteWsPaths = [] } = useWorkspaceContext();
  const items = useMemo(() => {
    return filterItems(wsName, query, noteWsPaths);
  }, [query, noteWsPaths, wsName]);

  const { getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
    () => false,
  );

  useEffect(() => {
    if (query === ']]' || query.includes('[[')) {
      replaceSuggestionMarkWith(palettePluginKey, '')(
        view.state,
        view.dispatch,
        view,
      );
    } else if (query.endsWith(']]')) {
      createBackLinkNode(wsPathFromQuery(query, wsName), noteWsPaths)(
        view.state,
        view.dispatch,
        view,
      );
    }
  }, [query, wsName, view, noteWsPaths]);

  return (
    <>
      {query === '' && (
        <InlinePaletteRow
          dataId="searchNote"
          className="palette-row"
          title={'ðŸ’¡ Search for a note to link it'}
        />
      )}
      {items.map((item, i) => {
        return (
          <InlinePaletteRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            description={item.title}
            rightHoverIcon={item.rightHoverIcon}
            rightIcon={
              <kbd className="whitespace-nowrap">{item.keybinding}</kbd>
            }
            {...getItemProps(item, i)}
          />
        );
      })}
    </>
  );
}

export function filterItems(wsName, query, allNoteWsPaths) {
  if (!query) {
    return [];
  }

  const items = allNoteWsPaths.map((wsPath) => {
    return {
      wsPath,
      uid: wsPath,
      title: removeMdExtension(resolvePath(wsPath).filePath),
      editorExecuteCommand: () => {
        return createBackLinkNode(wsPath, allNoteWsPaths);
      },
    };
  });

  let filteredItems = items
    .filter((item) => queryMatch(item, query))
    .slice(0, 15);

  const exactMatch = filteredItems.find(
    (item) =>
      item.title === query ||
      conditionalSuffix(query, '.md') === resolvePath(item.wsPath).filePath,
  );
  if (!exactMatch && query.length > 0) {
    const wsPath = wsPathFromQuery(query, wsName);

    filteredItems.unshift({
      uid: 'create-' + wsPath,
      title: 'Create: ' + removeMdExtension(resolvePath(wsPath).filePath),
      editorExecuteCommand: () => {
        return createBackLinkNode(wsPath, allNoteWsPaths);
      },
    });
  }

  return filteredItems;
}

function queryMatch(item, query) {
  const keywords = item.keywords || '';

  if (keywords.length > 0) {
    if (strMatch(keywords.split(','), query)) {
      return item;
    }
  }
  return strMatch(item.title, query) ? item : undefined;
}

function strMatch(a, b) {
  b = b.toLocaleLowerCase();
  if (Array.isArray(a)) {
    return a.filter(Boolean).some((str) => strMatch(str, b));
  }

  a = a.toLocaleLowerCase();
  return a.includes(b) || b.includes(a);
}

export function getBacklinkPath(wsPath, allWsPaths) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeMdExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeMdExtension(filePath);
  }

  return removeMdExtension(fileName);
}

function wsPathFromQuery(query, wsName) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = conditionalSuffix(filePath, '.md');
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}
