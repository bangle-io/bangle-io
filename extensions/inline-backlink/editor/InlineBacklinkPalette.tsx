import React, { useEffect, useMemo } from 'react';
import reactDOM from 'react-dom';

import type { EditorState, EditorView } from '@bangle.dev/pm';
import { useEditorViewContext } from '@bangle.dev/react';

import { nsmApi2 } from '@bangle.io/api';
import { byLengthAsc, useFzfSearch } from '@bangle.io/fzf-search';
import {
  replaceSuggestionMarkWith,
  useInlinePaletteItems,
  useInlinePaletteQuery,
} from '@bangle.io/inline-palette';
import { InlinePaletteRow, UniversalPalette } from '@bangle.io/ui-components';
import { assertNotUndefined, insertAt } from '@bangle.io/utils';
import {
  removeExtension,
  resolvePath,
  suffixWithNoteExtension,
} from '@bangle.io/ws-path';

import { backlinkNodeName, palettePluginKey } from '../config';
import { getBacklinkPath, wsPathFromQuery } from '../utils';

const FZF_SEARCH_LIMIT = 12;

// Creating this also closes the palette
const createBacklinkNode = (
  wsPath: string,
  allNoteWsPaths: readonly string[],
) => {
  return (
    state: EditorState,
    dispatch: EditorView['dispatch'] | undefined,
    view: EditorView | undefined,
  ) => {
    const nodeType = state.schema.nodes[backlinkNodeName];
    const backlinkPath = getBacklinkPath(wsPath, allNoteWsPaths);

    assertNotUndefined(nodeType, 'wikiLink must be defined');

    return replaceSuggestionMarkWith(
      palettePluginKey,
      nodeType.create({
        path: backlinkPath,
      }),
    )(state, dispatch, view);
  };
};

export function InlineBacklinkPalette() {
  const view = useEditorViewContext();

  if (!view || view.isDestroyed) {
    return null;
  }

  return <InlineBacklinkPalettePortal />;
}

function InlineBacklinkPalettePortal() {
  const { query, counter, tooltipContentDOM, isVisible } =
    useInlinePaletteQuery(palettePluginKey);

  return reactDOM.createPortal(
    <InlineBacklinkPaletteWrapper
      query={query}
      counter={counter}
      isVisible={isVisible}
      editorView={useEditorViewContext()}
    />,
    tooltipContentDOM,
  );
}

export function InlineBacklinkPaletteWrapper({
  query,
  counter,
  isVisible,
  editorView,
}: {
  query: string;
  counter: number;
  isVisible: boolean;
  editorView: EditorView;
}) {
  return (
    <div className="shadow-2xl B-ui-components_inline-palette-wrapper flex flex-col bg-colorNeutralBgLayerFloat">
      <div className="B-ui-components_inline-palette-items-wrapper">
        {/* TODO I am unable to hide inner when palette is invisible */}
        {isVisible && (
          <InlineBacklinkPaletteInner
            query={query}
            counter={counter}
            editorView={editorView}
          />
        )}
        <UniversalPalette.PaletteInfo>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">â†‘â†“</kbd> Navigate
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Enter</kbd> Create link
          </UniversalPalette.PaletteInfoItem>
          <UniversalPalette.PaletteInfoItem>
            <kbd className="font-normal">Esc</kbd> Dismiss
          </UniversalPalette.PaletteInfoItem>
        </UniversalPalette.PaletteInfo>
      </div>
    </div>
  );
}

const EMPTY_ARRAY: string[] = [];

function InlineBacklinkPaletteInner({
  query,
  editorView: view,
  counter,
}: {
  query: string;
  counter: number;
  editorView: EditorView;
}) {
  const { wsName, noteWsPaths = EMPTY_ARRAY } =
    nsmApi2.workspace.useWorkspace();

  const match = useFzfSearch(noteWsPaths, query, {
    limit: FZF_SEARCH_LIMIT,
    selector: (item: any) => resolvePath(item, true).filePath,
    tiebreakers: [byLengthAsc],
  });

  const items = useMemo(() => {
    let res = match.map((r) => {
      const wsPath = r.item;

      return {
        wsPath: wsPath,
        uid: wsPath,
        title: removeExtension(resolvePath(wsPath, true).filePath),
        editorExecuteCommand: () => {
          return createBacklinkNode(wsPath, noteWsPaths);
        },
      };
    });

    const exactMatch = res.find(
      (item) =>
        item.title === query ||
        suffixWithNoteExtension(query) ===
          resolvePath(item.wsPath, true).filePath,
    );

    if (!exactMatch && query.length > 0 && wsName) {
      const wsPath = wsPathFromQuery(query, wsName);

      let createItem = {
        uid: 'create-' + wsPath,
        wsPath: wsPath,
        title: 'Create: ' + removeExtension(resolvePath(wsPath).filePath),
        editorExecuteCommand: () => {
          return createBacklinkNode(wsPath, noteWsPaths);
        },
      };
      res = insertAt(res, 1, createItem);
    }

    return res;
  }, [noteWsPaths, match, wsName, query]);

  const { getItemProps } = useInlinePaletteItems(
    palettePluginKey,
    items,
    counter,
    () => false,
  );

  useEffect(() => {
    if (!wsName) {
      return;
    }
    if (query === ']]' || query.includes('[[')) {
      replaceSuggestionMarkWith(palettePluginKey, '')(
        view.state,
        view.dispatch,
        view,
      );
    } else if (query.endsWith(']]')) {
      createBacklinkNode(wsPathFromQuery(query, wsName), noteWsPaths)(
        view.state,
        view.dispatch,
        view,
      );
    }
  }, [query, wsName, view, noteWsPaths]);

  return (
    <>
      {query === '' && (
        <InlinePaletteRow
          dataId="searchNote"
          className="palette-row"
          title={'ðŸ’¡ Search for a note to link it'}
        />
      )}
      {items.map((item, i) => {
        return (
          <InlinePaletteRow
            key={item.uid}
            dataId={item.uid}
            className="palette-row"
            description={item.title}
            {...getItemProps(item, i)}
          />
        );
      })}
    </>
  );
}
