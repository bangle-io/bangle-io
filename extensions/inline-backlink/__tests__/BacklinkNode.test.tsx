/**
 * @jest-environment @bangle.io/jsdom-env
 */

import { setupTestExtension } from '@bangle.io/test-utils-2';
import React from 'react';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import inlineBackLinkExtension from '../index';
import { InlineBacklinkPaletteWrapper } from '../editor/InlineBacklinkPalette';
import { nsmApi2 } from '@bangle.io/api';
import { assertNotUndefined, sleep } from '@bangle.io/utils';
import { BacklinkWidget } from '../BacklinkWidget';

let abortController = new AbortController();

beforeEach(() => {
  abortController = new AbortController();
});

afterEach(async () => {
  abortController.abort();
});

async function setup({
  wsName,
  notes,
  fullEditor = false,
}: { wsName?: string; notes?: [string, string][]; fullEditor?: boolean } = {}) {
  const ctx = setupTestExtension({
    extensions: [inlineBackLinkExtension],
    abortSignal: abortController.signal,
    editor: true,
    fullEditor: fullEditor,
  });

  if (wsName) {
    await ctx.createWorkspace(wsName);
  }

  await ctx.createNotes(notes, { loadFirst: true });

  return ctx;
}

describe('BacklinkNode', () => {
  test('renders when no link found', async () => {
    const wsPath = 'test-ws:hi.md';
    const ctx = await setup({
      wsName: 'test-ws',
      notes: [[wsPath, `hello world`]],
    });

    await render(
      <ctx.ContextProvider>
        <BacklinkWidget />
      </ctx.ContextProvider>,
    );

    await sleep(50);

    expect(screen.getByTestId('inline-backlink_widget-container'))
      .toMatchInlineSnapshot(`
      <div
        class="flex flex-col"
        data-testid="inline-backlink_widget-container"
      >
        <span>
          üêí No backlinks found!
          <br />
          <span
            class="font-light"
          >
            Create one by typing 
            <kbd
              class="font-normal"
            >
              [[
            </kbd>
             followed by the name of the note.
          </span>
        </span>
      </div>
    `);
  });

  test('renders when no link found', async () => {
    const wsPath = 'test-ws:hi.md';
    const ctx = await setup({
      wsName: 'test-ws',
      notes: [[wsPath, `hello world [[my/note-path|monako]]`]],
      fullEditor: true,
    });

    await render(
      <ctx.ContextProvider>
        <BacklinkWidget />
      </ctx.ContextProvider>,
    );

    await sleep(50);

    expect(screen.getByTestId('inline-backlink-button')).toMatchInlineSnapshot(`
      <button
        aria-label="monako"
        class="B-inline-backlink_backlink-node hover:underline inline-flex gap-0_5 flex-row items-center rounded py-0 px-1 mx-1 text-start B-inline-backlink_backlink-node-not-found "
        data-testid="inline-backlink-button"
        draggable="false"
      >
        <span>
          <svg
            class="h-4 w-4 text-colorPromoteIcon"
            stroke="currentColor"
            style="fill: var(--BV-miscEditorBacklinkBg);"
            viewBox="0 0 18 18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10,5.5V1H3.5a.5.5,0,0,0-.5.5v15a.5.5,0,0,0,.5.5h11a.5.5,0,0,0,.5-.5V6H10.5A.5.5,0,0,1,10,5.5Z"
            />
            <path
              d="M11,1h.043a.5.5,0,0,1,.3535.1465l3.457,3.457A.5.5,0,0,1,15,4.957V5H11Z"
            />
          </svg>
        </span>
        <span
          class="inline whitespace-break-spaces"
        >
          monako
        </span>
      </button>
    `);
  });
});
