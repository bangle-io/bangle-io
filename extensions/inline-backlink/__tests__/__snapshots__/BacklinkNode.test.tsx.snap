// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BacklinkNode clicking no click if path validation fails 1`] = `
<div>
  <button
    class="b-inline-backlink_backlink b-inline-backlink_backlinkNotFound"
    draggable="false"
  >
    <svg
      class="inline-block"
      stroke="currentColor"
      viewBox="0 0 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10,5.5V1H3.5a.5.5,0,0,0-.5.5v15a.5.5,0,0,0,.5.5h11a.5.5,0,0,0,.5-.5V6H10.5A.5.5,0,0,1,10,5.5Z"
      />
      <path
        d="M11,1h.043a.5.5,0,0,1,.3535.1465l3.457,3.457A.5.5,0,0,1,15,4.957V5H11Z"
      />
    </svg>
    <span
      class="inline"
    >
      Invalid link (monako)
    </span>
  </button>
</div>
`;
